### PR TITLE
GH-36612: [C++] Ensure compatibility between std::span and arrow::util::span 

### PR DESCRIPTION
**Pull request by Divyansh200102**
* Closes: #36612

### CODE REVIEW

**Comment by llama90** [comment]
Hello. Thank you for your contribution.

However, when considering submitting a PR to Arrow, it would be advisable to adhere to the existing template provided.

**Comment by Divyansh200102** [comment]
> Hello. Thank you for your contribution.
> 
> However, when considering submitting a PR to Arrow, it would be advisable to adhere to the existing template provided.

Can you please tell what is causing checks to fail and what I can i do to correct them or am I doing something wrong?

File: cpp/src/arrow/util/span.h [hunk 1]
```diff
@@ -129,4 +129,72 @@ constexpr span<std::byte> as_writable_bytes(span<T> s) {
   return {reinterpret_cast<std::byte*>(s.data()), s.size_bytes()};
 }
 
+template <>
+class arrow::util::span<std::byte> : public std::span<std::byte> {
```
Comment (llama90) [inline cpp/src/arrow/util/span.h:133]: I copied your code and was able to identify the following issues.

`std::span` to be `span`

Also, `arrow::util::span<std::byte>` case, `Explicit specialization of 'arrow::util::span<std::byte>' after instantiation implicit instantiation first required here.`

File: cpp/src/arrow/util/span.h [hunk 2]
```diff
@@ -129,4 +129,72 @@ constexpr span<std::byte> as_writable_bytes(span<T> s) {
...
+    for (size_t i = 0; i < size_; ++i) {
+      data_[i] = arrow::util::span<std::byte>(other[i].data(), other[i].size());
+    }
+  }
+
+  arrow::util::span<std::byte>& operator[](size_t idx) { return data_[idx]; }
+
+  const arrow::util::span<std::byte>& operator[](size_t idx) const { return data_[idx]; }
+
+  arrow::util::span<std::byte>* data() const { return &data_[0]; }
```
Comment (llama90) [inline cpp/src/arrow/util/span.h:170]: `&data_[0];` case, `Cannot initialize return object of type 'arrow::util::span<std::byte> *' with an rvalue of type 'const std::vector<arrow::util::span<std::byte>>::value_type *' (aka 'const arrow::util::span<std::byte> *')`

**Review by llama90 — COMMENTED** [review]
In my opinion, it seems like you are only adding code without performing tasks such as building or testing.

I believe that contributing code begins with understanding and setting up the development environment.

If you haven't set up your development environment yet, it would be beneficial to start there, and you can refer to the following [document](https://arrow.apache.org/docs/developers/cpp/building.html) for guidance.

**Comment by Divyansh200102** [comment]
> In my opinion, it seems like you are only adding code without performing tasks such as building or testing.
> 
> I believe that contributing code begins with understanding and setting up the development environment.
> 
> If you haven't set up your development environment yet, it would be beneficial to start there, and you can refer to the following [document](https://arrow.apache.org/docs/developers/cpp/building.html) for guidance.

Thanks for replying @llama90 I will try and address the things you told me to for now I will set this pr as draft.

**Pull request closed without merging**